clash-lib/prims/verilog/Clash_Sized_Vector.json

```diff
@@ -123,7 +123,7 @@ end
     { "name"      : "Clash.Sized.Vector.imap"
     , "workInfo"  : "Never"
     , "kind"      : "Declaration"
-    , "type"      : "imap :: KnownNat n => (Index n -> a -> b) -> Vec n a -> Vec n b"
+    , "type"      : "imap :: forall sat n a b . (KnownNat n, 1 <= n) => (SatIndex sat n -> a -> b) -> Vec n a -> Vec n b"
```

Comment:
Why is this constraint added? Seems like it should work and just return an empty vector?

Reply:
Also: an extra constraint is an extra argument in Core, so you'd need to change the blackbox implementation too unfortunately..

Reply:
The `KnownNat` constraint is added because `imap` uses the `Num` instance which now uses the `SaturatingNum` instance which has that constraint. An `Index 0` causes all sorts of problems since it doesn't have any valid value.

Reply:
What kind of issues are you thinking of? HDL generation should filter it right out, so I don't think you have to consider that.

Reply:
`Index 1` has a valid value of just 0. `Index 0` has no valid value ever. It does get filtered out when compiling to HDL but any evaluation of a value of `Index 0` leads to an error.

Reply:
I can try to look into removing the `1 <= n` constraint from `SaturatingNum`.

Reply:
Yeah, just removing it should be fine. Like you noted, we can't express things such as `imap` sensibly.

Reply:
Yesterday evening I have unsuccesfully tried various things to be able to remove that constraint. The problem lies in the `AResult` and `MResult` types which subtract one from the original width in the calculation of the new width. I had hoped I could do something `(Max n 1) - 1` to  ensure it would always be one but the constraint solver doesn't pick up on this.

Do you have any pointers how I could achieve this?

Reply:
I have looked into the `GHC.TypeLits.Extra` package as I thought I could relatively easily resolve `a <= Max a b` and `b <= Max a b` based on the existence of `Max`. However the solver in that package already should handle that case. Since when the following added test case is handled correctly by the package.

```patch
diff --git a/tests/Main.hs b/tests/Main.hs
index 443a2dd..f4368e5 100644
--- a/tests/Main.hs
+++ b/tests/Main.hs
@@ -18,6 +18,12 @@ import ErrorTests
 import GHC.TypeLits
 import GHC.TypeLits.Extra

+f :: (1 <= n) => Proxy n -> Proxy n
+f = id
+
+test0 :: Proxy (Max n 1) -> Proxy (Max n 1)
+test0 = f
+
 test1 :: Proxy (GCD 6 8) -> Proxy 2
 test1 = id

@@ -165,7 +171,10 @@ main = defaultMain tests
 tests :: TestTree
 tests = testGroup "ghc-typelits-natnormalise"
   [ testGroup "Basic functionality"
-    [ testCase "GCD 6 8 ~ 2" $
+    [ testCase "1 <= Max n 1" $
+      show (test0 Proxy) @?=
+      "Proxy"
+    , testCase "GCD 6 8 ~ 2" $
       show (test1 Proxy) @?=
       "Proxy"
     , testCase "forall x . GCD 6 8 + x ~ x + GCD 10 8" $
```

However pasting the `f` and `test0` into a file in the clash-prelude causes the compilation to fail:

```patch
diff --git a/clash-prelude/src/Clash/Class/Num.hs b/clash-prelude/src/Clash/Class/Num.hs
index 376e1e62..b5f22d59 100644
--- a/clash-prelude/src/Clash/Class/Num.hs
+++ b/clash-prelude/src/Clash/Class/Num.hs
@@ -15,6 +15,12 @@ Maintainer :  Christiaan Baaij <christiaan.baaij@gmail.com>

 {-# OPTIONS_HADDOCK show-extensions #-}

+
+{-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Normalise       #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Normalise       #-}
+
 module Clash.Class.Num
   ( -- * Arithmetic functions for arguments and results of different precision
     ExtendingNum (..)
@@ -27,10 +33,14 @@ module Clash.Class.Num
   , boundedMul
   , satSucc
   , satPred
+  , test0
   )
 where

-import GHC.TypeLits               (TypeError, ErrorMessage (Text))
+import GHC.TypeLits               (type (<=), TypeError, ErrorMessage (Text))
+
+import GHC.TypeLits.Extra         (Max)
+import Data.Proxy (Proxy (..))

 -- * Arithmetic functions for arguments and results of different precision

@@ -71,6 +81,12 @@ data SaturationMode
               -- cost with the index type.
   deriving Eq

+f :: (1 <= n) => Proxy n -> Proxy n
+f = id
+
+test0 :: Proxy (Max n 1) -> Proxy (Max n 1)
+test0 = f
+

 class KnownSatMode (sat :: SaturationMode) where
   satMode :: SaturationMode
```

with the following error message:

```
[ 9 of 89] Compiling Clash.Class.Num  ( src/Clash/Class/Num.hs, dist/build/Clash/Class/Num.o )

src/Clash/Class/Num.hs:88:9: error:
    • Couldn't match type ‘1 GHC.TypeNats.<=? Max n 1’ with ‘'True’
        arising from a use of ‘f’
    • In the expression: f
      In an equation for ‘test0’: test0 = f
    • Relevant bindings include
        test0 :: Proxy (Max n 1) -> Proxy (Max n 1)
          (bound at src/Clash/Class/Num.hs:88:1)
   |
88 | test0 = f
```

Reply:
I got it to work. I forgot to load the necessary GHC plugin...

Reply:
Sorry I couldn't look at it sooner.

You could add `{-# OPTIONS_GHC -fconstraint-solver-iterations=5 #-}` locally no? Or is it triggered at use sites?

Reply:
I first believed It was triggered at use sites. But some testing I can't trigger it from a use site. doctests was complaining about it so I assumed use sites where the problem.

Reply:
Could you take a look at this? Doctest is still failing due to `constraint-solver-iteration`. I have even added `-fconstraint-solver-iterations=5` to the doctest entry in the `clash-prelude.cabal` file but it still reports the error.

Reply:
I'll have a look

Reply:
We're seeing another issue when we compile locally :-/. The type checker plugins can't figure out that `Max (n+1) 1 ~ (n+1)` . We suspect normalization (turning `n+1` into `1+n`) doesn't work inside `Max`. We're on the case @rowanG077, we'll hopefully fix it after the weekend..

Reply:
This is weird. I don't have this issue locally nor does CI have this issue. Where in the build process does this occur?

Reply:
Did you get the `solveSimpleWanteds: too many iterations` errors locally just like the CI?

Martijn and I only got a lot of type errors during doctests of `Clash.Sized.Vector` and `Clash.Sized.RTree`

Reply:
No I got different errors also due to constraint solving. I hacked the default.nix to enable the testing so I should get the correct libs. I assumed that it was due to my setup not being correct and used the CI as a source of truth for doctests.

```
nix-shell:~/engineering/clash-compiler/clash-prelude]$ cabal doctest
Warning: The doctest command is a part of the legacy v1 style of cabal usage.

Please switch to using either the new project style and the new-doctest
command or the legacy v1-doctest alias as new-style projects will become the
default in the next version of cabal-install. Please file a bug if you cannot
replicate a working v1- use case with the new-style commands.

For more information, see: https://wiki.haskell.org/Cabal/NewBuild

Preprocessing library for clash-prelude-0.9999..

src/Clash/Class/Resize.hs:35:12: error:
    • Could not deduce (KnownNat (b + a))
        arising from a use of ‘resize’
      from the context: Resize f
        bound by the class declaration for ‘Resize’
        at src/Clash/Class/Resize.hs:21:7-12
      or from: (KnownNat a, KnownNat b)
        bound by the type signature for:
                   extend :: forall (a :: Nat) (b :: Nat).
                             (KnownNat a, KnownNat b) =>
                             f a -> f (b + a)
        at src/Clash/Class/Resize.hs:34:13-56
    • In the expression: resize
      In an equation for ‘extend’: extend = resize
   |
35 |   extend = resize
   |            ^^^^^^

src/Clash/Class/Resize.hs:40:16: error:
    • Could not deduce (KnownNat (b + a))
        arising from a use of ‘resize’
      from the context: Resize f
        bound by the class declaration for ‘Resize’
        at src/Clash/Class/Resize.hs:21:7-12
      or from: (KnownNat a, KnownNat b)
        bound by the type signature for:
                   signExtend :: forall (a :: Nat) (b :: Nat).
                                 (KnownNat a, KnownNat b) =>
                                 f a -> f (b + a)
        at src/Clash/Class/Resize.hs:39:17-60
    • In the expression: resize
      In an equation for ‘signExtend’: signExtend = resize
   |
40 |   signExtend = resize
```

Reply:
Did you guys have time to look at this? I have seen that a solver has been updated.

Reply:
Yeah! So, `ghc-typelits-extra` can now solve `Max (n+1) 1 ~ (1+n)` which should solve the problems we were seeing. I'll give it another spin as I'm waiting on other compiles anyway..

Reply:
It does seem to solve that now, but it gets stuck on another error:

```
 but got: 
          <interactive>:4091:45: error:
              • Couldn't match type ‘1 <=? Max (((2 ^ l) + 1) + ((2 ^ l) + 1)) 1’
                               with ‘'True’
                Expected type: IIndex @@ (l + 1)
                  Actual type: AResult
                                 (SatIndex 'SatError ((2 ^ l) + 1))
                                 (SatIndex 'SatError ((2 ^ l) + 1))
              • In the expression: add x y
                In the third argument of ‘dtfold’, namely ‘(\ _ x y -> add x y)’
                In the expression:
                  dtfold (Proxy @IIndex) fromIntegral (\ _ x y -> add x y) (bv2v bv)
              • Relevant bindings include
                  y :: IIndex @@ l (bound at <interactive>:4091:40)
                  x :: IIndex @@ l (bound at <interactive>:4091:38)

```

The type checkers are able to solve that though:

https://github.com/clash-lang/ghc-typelits-extra/blame/master/src/GHC/TypeLits/Extra/Solver.hs#L156-L157

..so I'm not sure what's up.

Reply:
It could be that the solver are loaded in a wrong order at the top of the file?

Reply:
I believe they run in a "loop", so I don't think that's it..